### ISSUE TITLE
ModelMetadata.EnumDisplayNamesAndValues explicitly skips [Display(Name = string.Empty)]

### ISSUE THREAD

**Issue by Martijn-Sharp** [aspnet/Mvc#2851]
When defining an enum with [Display] attributes in it, there's no way to use [Display(Name = string.Empty)] (on for example the first item, due to requirements)
This is because of what happens on line 237 in method GetDisplayName: https://github.com/aspnet/Mvc/blob/e985fa5d423968f36a651a354d9c953192c52532/src/Microsoft.AspNet.Mvc.DataAnnotations/DataAnnotationsMetadataProvider.cs

The reasoning is absent, but I *guess* it's because when the Name property isn't set then the display.GetName() shouldn't be returned at all.

That's why I think the check probably shouldn't be '!string.IsEmptyOrNull(name)' but rather just a null check, because as I pointed out before, assigning a string.Empty might be a requirement.
For now I'm using a space to circumvent this ([Display(Name = " ")])

**Comment by dougbu** [comment]
> as I pointed out before, assigning a string.Empty might be a requirement

Think I missed this. Why might an empty display name be a requirement?

I'm not asking whether this is a bug or not: It is since the BCL doesn't special-case empty strings for `[Display(Name=...)]`. Just curious about your scenario.

**Comment by Martijn-Sharp** [comment]
My scenario is as following:

To easily make a html select-element of an enum I made a custom taghelper (SelectEnumTagHelper) which converts an enum into a select-list. This is where the ModelExpression.ModelMetadata.EnumDisplayValuesAndNames came in. I'm working on a web application which has some forms on it, on one of the forms is a html select-element which contains some values. In the model in the MVC framework I have it defined the property and enum as following (fake values coming up):
public FakeOptions MyProperty { get; set; }

public enum FakeOptions
{
[Display(Name = "")]
None,
[Display(Name = "Minions")]
Banana,
Purple,
Kepler
}

And in the view:
<select-enum asp-for="MyProperty"></select-enum>

The 'None' is in it because I'd like to know, in the MVC framework, whether someone has selected 'none'. But a requirement from one of my 'gui guys' is that in the front-end there's no text whatsoever in that option.

I could circumvent this in a few ways, perhaps using javascript, or some other way of adding a text-less select-option element. I could also write my own enum-to-selectlist method. But ideally I'd do it the way I described before, with assigning an empty string to an enum value, using existing framework methods. So that an enum gets resolved automatically on the front-end based fully on the enum I defined in MVC.
